add openjdk11

grueni opened the pull request:
I used the patches from OmniosCe and Peter Tribblix.

If openjdk11 should be defined as latest the following packages must be changed.
developer/java/openjdk8
runtime/java/jruby
service/resource-pools/poold

The former runtime in JRE disappeared with openjdk9.

grueni committed:
openjdk11: initial commit

AndWac commented:
It doesn't build here:
The following warnings were produced. Repeated here for convenience:
WARNING: Value for VERSION_BUILD has been sanitized from '9.1' to '9'
WARNING: Ignoring CFLAGS(-m64 -O3 ) found in environment. Use --with-extra-cflags
WARNING: Ignoring CXXFLAGS( -m64 -O3 ) found in environment. Use --with-extra-cxxflags
WARNING: Ignoring LDFLAGS(-m64) found in environment. Use --with-extra-ldflags
WARNING: You are using a linker older than 2.18. This is not a supported configuration.
WARNING: The following variables might be unknown to configure:  CC  CXX  F77  FC  CFLAGS  CXXFLAGS  FFLAGS  FCFLAGS  LDFLAGS  PKG_CONFIG_PATH

(cd /oibuild/oi-userland/components/runtime/openjdk-11/build/amd64; /bin/cp -f -r build/* .;)
/usr/bin/touch /oibuild/oi-userland/components/runtime/openjdk-11/build/amd64/.configured
(cd /oibuild/oi-userland/components/runtime/openjdk-11/build/amd64 ; /usr/bin/env LD_OPTIONS="-M /usr/lib/ld/map.noexstk -M /usr/lib/ld/map.noexdata -M /usr/lib/ld/map.pagealign -Bdirect -z ignore" LD_EXEC_OPTIONS="-z aslr=disable"  "ac_cv_func_realloc_0_nonnull=yes" \
	/usr/gnu/bin/make  -j4  \
	images)
make[1]: Entering directory '/oibuild/oi-userland/components/runtime/openjdk-11/build/amd64'
make[2]: Entering directory '/oibuild/oi-userland/components/runtime/openjdk-11/openjdk-jdk11u-jdk-11.0.10-9-1'
make[2]: Leaving directory '/oibuild/oi-userland/components/runtime/openjdk-11/openjdk-jdk11u-jdk-11.0.10-9-1'
Error: 'make -jN' is not supported, use 'make JOBS=N'
/oibuild/oi-userland/components/runtime/openjdk-11/openjdk-jdk11u-jdk-11.0.10-9-1/make/Init.gmk:115: *** Cannot continue.  Stop.
make[1]: Leaving directory '/oibuild/oi-userland/components/runtime/openjdk-11/build/amd64'
gmake: *** [/oibuild/oi-userland/make-rules/configure.mk:183: /oibuild/oi-userland/components/runtime/openjdk-11/build/amd64/.built] Error 2

grueni commented:
Don't use -j4 with make.
Number of cores and ram are probed during the compilation.
If you would like to restrict the ressources used by the compilation use the options in the Makefile
  --with-num-cores        number of cores in the build system, e.g.
                          --with-num-cores=8 [probed]
  --with-memory-size      memory (in MB) available in the build system, e.g.
                          --with-memory-size=1024 [probed]

AndWac commented:
On our build server the jenkins job sets -j8. You'll need to override this explicitly in the Makefile with
COMPONENT_BUILD_ARGS=

AndWac commented on components/runtime/openjdk-11/Makefile:
```diff
@@ -0,0 +1,146 @@
+REQUIRED_PACKAGES += x11/library/libxtst
+REQUIRED_PACKAGES += system/library/c++/sunpro
+REQUIRED_PACKAGES += system/header/header-audio
+
```
If I run gmake REQUIRED_PACKAGES here, the last two entries aren't detected. Is library/c++/sunpro really needed?

grueni commented:
Will check it.
At least system/header/header-audio is required if audio should be supported.

AndWac commented:
Then you can move the header/header-audio line above the # Auto-generated dependencies and comment it with something like # Manually added build dependencies

AndWac commented on components/runtime/openjdk-11/openjdk.p5m:
```diff
@@ -0,0 +1,1011 @@
+## as long as the two links are defined in openjdk8 they can't used here
+# link path=usr/jdk/latest target=$(OPENJDK_INSTANCE) pkg.linted.userland.action002.0=true
+# link path=usr/java target=jdk/$(OPENJDK_INSTANCE)
+
```
The way to handle this would normally be to create a separate PR which removes this from openjdk-8 (increasing its COMPONENT_REVISION) and add a comment to the PR saying "Merge this PR after the one for openjdk-8".

AndWac replied:
Thinking more about the latest links makes me wonder whether it should also be mediated.

grueni replied:
Good idea to mediate the two links

pyhalov replied:
Not sure. Latest clearly should be 'latest', '/usr/java' once was used by illumos-gate builds. Perhaps, now it can be mediated.

grueni replied:
It depends on the definition of "latest". From the view point of the repository it should it should be linked to the newest available version. And not be mediated.

AndWac commented:
pkg5 file seems to miss. You can generate it either by running gmake publish or by gmake update-metadata.

AndWac commented:
I have build and installed it successfully. When I set the java mediator to 11 it runs tomcat fine. I still need to check jenkins.
But when I try to rebuild the package with openjdk11 as default java it fails with
configure: WARNING: Value for VERSION_BUILD has been sanitized from '9.1' to '9'
checking for version string... 11.0.10+9-adhoc.andreas.openjdk-jdk11u-jdk-11.0.10-9-1
configure: Found potential Boot JDK using configure arguments
configure: Potential Boot JDK found at /oibuild/oi-userland/components/runtime/openjdk-11/jdk11.0.6 is incorrect JDK version (Error: dl failure on line 597); ignoring
configure: (Your Boot JDK version must be one of: 10 11)
configure: error: The path given by --with-boot-jdk does not contain a valid Boot JDK
configure exiting with result code 1
gmake: *** [/oibuild/oi-userland/make-rules/configure.mk:175: /oibuild/oi-userland/components/runtime/openjdk-11/build/amd64/.configured] Error 1

AndWac commented:
jenkins has a problem. It thinks it is offline and wants to configure a proxy.
When I try to configure plugins I'll get a warning:
There were errors checking the update sites: SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty

This works with openjdk-8. So I think there's a problem with the SSL or certificate handling in openjdk-11.

grueni commented:
The default location for cacerts in openjdk8 is in jre/lib/security and in openjdk11 in lib/security.
Both versions can read their cacerts.
in openjdk8:
/usr/jdk/instances/openjdk1.8.0/bin/keytool -list -storepass changeit -keystore /usr/jdk/instances/openjdk1.8.0/jre/lib/security/cacerts |less
in openjdk11:
/usr/jdk/instances/openjdk11.0.10/bin/keytool -list -storepass changeit -cacerts |less
or
/usr/jdk/instances/openjdk11.0.10/bin/keytool -list -storepass changeit -keystore /usr/jdk/instances/openjdk11.0.10/lib/security/cacerts 2>&1 |less

grueni commented:
> configure: Potential Boot JDK found at /oibuild/oi-userland/components/runtime/openjdk-11/jdk11.0.6 is incorrect JDK version (Error: dl failure on line 597); ignoring

Still investigating this problem.
I guess this problem must be solved at first.

grueni committed:
added pkg5, marked manually added REQUIRED-PACKAGES, ignore -j8 from Jenkins

AndWac commented:
When system/library/c++/sunpro is installed jenkins works as expected. I had it installed when I created the package but uninstalled it when testing tomcat8 and jenkins. Now I will build a package without sunpro installed and hope to find out whether that works.
Either way we either need to make sunpro a runtime requirement or make sure it's not installed when the package is being build.

AndWac commented:
I cannot build the package anymore even after uninstalling openjdk11. It always complains about the wrong java version. It seems to find itself when configured again.

grueni commented:
I built jenkins with sunpro installed successfully.
Now I am preparing an openjdk11 without sunpro installed and try it again.
I need to see the failures if any.

Without sunpro installed and with mediator set to 11 I get errors compiling openjdk11:
configure: WARNING: Value for VERSION_BUILD has been sanitized from '9.1' to '9'
checking for version string... 11.0.10+9-adhoc..openjdk-jdk11u-jdk-11.0.10-9-1
configure: Found potential Boot JDK using configure arguments
configure: Potential Boot JDK found at /export/builds/oi-userland/components/runtime/openjdk-11/jdk11.0.6 is incorrect JDK version (Error: dl failure on line 597); ignoring
configure: (Your Boot JDK version must be one of: 10 11)
configure: error: The path given by --with-boot-jdk does not contain a valid Boot JDK
configure exiting with result code 1
gmake: *** [/export/builds/oi-userland/make-rules/configure.mk:175: /export/builds/oi-userland/components/runtime/openjdk-11/build/amd64/.configured] Error 1

Will check it.

AndWac commented:
I didn't try to build it, I tried to run it: "pfexec svcadm enable jenkins".
Then I logged into it and started to configure it.

grueni commented:
If sunpro is not installed openjdk11 can't be built independent from the mediator for java. I tried both versions 8 and 11.
Maybe this is caused by the jdk used for bootstrapping: jdk11.0.6 from P. Tribble.
Will check it tomorrow.

AndWac commented:
Good catch. You'll need to re-add its build dependency (also above the auto-generated dependencies) and add a runtime dependency in the openjdk.p5m, eg. something like this:
\#  force a dependency on system/library/c++/sunpro as cryptography depends on it:
depend type=require fmri=system/library/c++/sunpro

grueni commented:
If system/library/c++/sunpro is not installed you get the errors concerning the worng version.
The real problem is a missing library for the java used for bootstrapping.

The feature detection of java version in make/autoconf/boot-jdk.m4 uses '../java -version' and this results in
root@build3:/export/builds/oi-userland/components/runtime/openjdk-11# jdk11.0.6/bin/java -version
Error: dl failure on line 597
Error: failed /export/builds/oi-userland/components/runtime/openjdk-11/jdk11.0.6/lib/server/libjvm.so, because ld.so.1: java: fatal: libdemangle.so.1: open failed: No such file or directory
and libdemangle.so.1 is part of system/library/c++/sunpro

root@build3:/export/builds/oi-userland/components/runtime/openjdk-11# pkg search -r libdemangle.so.1
INDEX      ACTION VALUE                          PACKAGE
basename   file   usr/lib/amd64/libdemangle.so.1 pkg:/system/library/c++/sunpro@0.5.11-2013.0.0.0
basename   file   usr/lib/libdemangle.so.1       pkg:/system/library/c++/sunpro@0.5.11-2013.0.0.0

For now the package system/library/c++/sunpro is a build dependency but not a runtime dependency.
The next time the package openjdk11 can be used for bootstrapping and the runtime dependency for system/library/c++/sunpro can be removed.

I used the new package openjdk11 for bootstrapping and built it again without an installed sunpro and the package was built without any problems.

grueni committed:
Makefile reads now -j from arguments for make
spawnhelper must be executable
system/library/c++/sunpro as runtime dependency
patches for flags-ldflags combined in one patch

AndWac commented:
Your latest changes made it necessary to re-run gmake REQUIRED_PACKAGES on my system. Running it detects two new dependencies:
REQUIRED_PACKAGES += image/library/libjpeg6
REQUIRED_PACKAGES += image/library/libjpeg6-ijg

Without them gmake publish barfed:
/oibuild/oi-userland/components/runtime/openjdk-11/build/manifest-i386-openjdk.depend has unresolved dependency '
    depend type=require fmri=__TBD pkg.debug.depend.file=libjpeg.so.62 \
        pkg.debug.depend.reason=usr/jdk/instances/openjdk11.0.10/lib/libsplashscreen.so \
        pkg.debug.depend.type=elf \
        pkg.debug.depend.path=lib/64 \
        pkg.debug.depend.path=usr/gcc/7/lib/amd64 \
        pkg.debug.depend.path=usr/java/openjdk11/lib \
        pkg.debug.depend.path=usr/java/openjdk11/lib/jli \
        pkg.debug.depend.path=usr/java/openjdk11/lib/server \
        pkg.debug.depend.path=usr/jdk/instances/openjdk11.0.10/lib \
        pkg.debug.depend.path=usr/lib/64 \
        pkg.debug.depend.path=usr/lib/libjpeg8-turbo/lib/amd64'.
gmake: *** [/oibuild/oi-userland/make-rules/ips.mk:493: /oibuild/oi-userland/components/runtime/openjdk-11/build/.resolved-i386] Error 1

AndWac commented:
> The default location for cacerts in openjdk8 is in jre/lib/security and in openjdk11 in lib/security.
> Both versions can read their cacerts.
> in openjdk8:
> /usr/jdk/instances/openjdk1.8.0/bin/keytool -list -storepass changeit -keystore /usr/jdk/instances/openjdk1.8.0/jre/lib/security/cacerts |less
> in openjdk11:
> /usr/jdk/instances/openjdk11.0.10/bin/keytool -list -storepass changeit -cacerts |less
> or
> /usr/jdk/instances/openjdk11.0.10/bin/keytool -list -storepass changeit -keystore /usr/jdk/instances/openjdk11.0.10/lib/security/cacerts 2>&1 |less

We need to find a solution such that jenkins will work with either jdk version.

grueni committed:
openjdk11: add additional REQUIRED_PACKAGES

grueni commented:
The truststore cacerts is in the wrong format. 
Java changed the defaults for creating keystores from JKS to PKSC12.
I will deliver a corrected version in the evening.

AndWac commented:
Am 04.05.21 um 14:38 schrieb grueni:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In components/runtime/openjdk-11/openjdk.p5m
> <https://github.com/OpenIndiana/oi-userland/pull/6716#discussion_r625744163>:
>
> > +set name=pkg.fmri	value=pkg:/$(COMPONENT_FMRI)@$(IPS_COMPONENT_VERSION),$(BUILD_VERSION)
> +set name=pkg.summary value="OpenJDK $(COMPONENT_VERSION) Runtime Environment"
> +set name=info.classification value="$(COMPONENT_CLASSIFICATION)"
> +set name=info.source-url value=$(COMPONENT_ARCHIVE_URL)
> +set name=info.upstream-url value=$(COMPONENT_PROJECT_URL)
> +set name=org.opensolaris.consolidation value=$(CONSOLIDATION)
> +
> +license $(COMPONENT_LICENSE_FILE) license='$(COMPONENT_LICENSE)'
> +
> +dir path=usr/jdk/instances/$(OPENJDK_INSTANCE)
> +link path=usr/jdk/$(OPENJDK_INSTANCE) target=instances/$(OPENJDK_INSTANCE)
> +
> +## as long as the two links are defined in openjdk8 they can't used here
> +# link path=usr/jdk/latest target=$(OPENJDK_INSTANCE) pkg.linted.userland.action002.0=true
> +# link path=usr/java target=jdk/$(OPENJDK_INSTANCE)
> +
>
> It depends on the definition of "latest". From the view point of the
> repository it should it should be linked to the newest available
> version. And not be mediated.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/pull/6716#discussion_r625744163>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADET4N73PXSCLKQEUIMECQDTL7TD7ANCNFSM437N7ZJQ>.
>
But that would mean that installing the latest version is mandatory. In
my eyes latest should be the latest available version but you are right,
a mediator wouldn't guarantee that, too.

grueni committed:
openjdk11: storetype=jks for caerts, added dependencies to pkg5

grueni commented:
Jenkins starts now without problems.

When usr/java will be changed, this packages must be checked:

../../image/plplot/Makefile:53:CMAKE_OPTIONS += -DJAVA_AWT_LIBRARY=/usr/java/jre/lib/amd64/libjawt.so
../../image/plplot/Makefile:54:CMAKE_OPTIONS += -DJAVA_JVM_LIBRARY=/usr/java/jre/lib/amd64/server/libjvm.so
../../database/mariadb-101/Makefile:87:CMAKE_OPTIONS.64+= -DJAVA_AWT_LIBRARY=/usr/java/jre/lib/amd64/libjawt.so
../../database/mariadb-101/Makefile:88:CMAKE_OPTIONS.64+= -DJAVA_JVM_LIBRARY=/usr/java/jre/lib/amd64/server/libjvm.s0
../../database/mariadb-103/Makefile:87:CMAKE_OPTIONS+= -DJAVA_AWT_LIBRARY=/usr/java/jre/lib/amd64/libjawt.so
../../database/mariadb-103/Makefile:88:CMAKE_OPTIONS+= -DJAVA_JVM_LIBRARY=/usr/java/jre/lib/amd64/server/libjvm.so

1) The packages miss the dependency on java
2) jre disappeared
3) The version of java is undefined in the packages

Independent of the change of usr/java the 3 packages should be changed according to the other packages where dependencies on openjkd7 oder openjkd8 are defined.
